File: ios/scribeai/ErrorReportingService.swift
import Foundation
import UIKit

/// Service for reporting user journey errors to the backend
class ErrorReportingService {
    static let shared = ErrorReportingService()

    private init() {}

    /// User journey flows that can be reported
    enum UserFlow: String {
        // Auth flows
        case signIn = "Sign In"
        case signUp = "Sign Up"
        case signInWithApple = "Sign In with Apple"
        case signInWithGoogle = "Sign In with Google"
        case signOut = "Sign Out"

        // Content creation flows
        case recordAudio = "Record Audio"
        case uploadPDF = "Upload PDF"
        case scanDocument = "Scan Document"
        case youtubeLink = "YouTube Link"
        case uploadSlideshow = "Upload Slideshow"

        // AI generation flows
        case generateSummary = "Generate Summary"
        case generateQuiz = "Generate Quiz"
        case generateFlashcards = "Generate Flashcards"
        case generatePodcast = "Generate Podcast"
        case generateDiagram = "Generate Diagram"
        case chatWithNote = "Chat with Note"

        // Subscription flows
        case purchase = "Purchase Subscription"
        case restorePurchases = "Restore Purchases"

        // Other flows
        case fetchNotes = "Fetch Notes"
        case deleteNote = "Delete Note"
        case transcription = "Transcription"
    }

    /// Report an error that occurred during a user journey
    /// Runs completely in background - never blocks UI
    /// - Parameters:
    ///   - flow: The user flow where the error occurred
    ///   - error: The error that occurred
    ///   - additionalInfo: Any additional context
    func reportError(flow: UserFlow, error: Error, additionalInfo: String? = nil) {
        // Fire and forget - detached task with low priority
        Task.detached(priority: .utility) { [weak self] in
            await self?.reportErrorAsync(flow: flow, error: error, additionalInfo: additionalInfo)
        }
    }

    /// Report an error with a custom error message
    /// Runs completely in background - never blocks UI
    /// - Parameters:
    ///   - flow: The user flow where the error occurred
    ///   - errorMessage: Custom error message
    ///   - additionalInfo: Any additional context
    func reportError(flow: UserFlow, errorMessage: String, additionalInfo: String? = nil) {
        // Fire and forget - detached task with low priority
        Task.detached(priority: .utility) { [weak self] in
            await self?.reportErrorAsync(flow: flow, errorMessage: errorMessage, additionalInfo: additionalInfo)
        }
    }

    private func reportErrorAsync(flow: UserFlow, error: Error, additionalInfo: String? = nil) async {
        var errorMessage = error.localizedDescription
        if let additionalInfo = additionalInfo {
            errorMessage += " | Additional: \(additionalInfo)"
        }
        await reportErrorAsync(flow: flow, errorMessage: errorMessage, additionalInfo: nil)
    }

    private func reportErrorAsync(flow: UserFlow, errorMessage: String, additionalInfo: String? = nil) async {
        let url = URL(string: "\(Constants.baseURL)/api/alerts/error")!

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10 // Short timeout - don't wait too long

        // Add auth token if available
        if let token = KeychainService.shared.get(Constants.Keychain.accessToken) {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Build error message
        var fullErrorMessage = errorMessage
        if let additionalInfo = additionalInfo {
            fullErrorMessage += " | \(additionalInfo)"
        }

        // Get device info
        let deviceInfo: [String: Any] = [
            "device": UIDevice.current.model,
            "osVersion": UIDevice.current.systemVersion,
            "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "Unknown"
        ]

        let body: [String: Any] = [
            "flow": flow.rawValue,
            "error": fullErrorMessage,
            "deviceInfo": deviceInfo
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)

            let (_, response) = try await URLSession.shared.data(for: request)

            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                print("📧 Error reported successfully for flow: \(flow.rawValue)")
            } else {
                print("⚠️ Failed to report error, status: \((response as? HTTPURLResponse)?.statusCode ?? -1)")
            }
        } catch {
            // Don't report errors about reporting errors - just log locally
            print("⚠️ Failed to send error report: \(error.localizedDescription)")
        }
    }
}

File: android/app/src/main/java/com/kreativekoala/scribeai/utils/ErrorReportingService.kt
package com.kreativekoala.scribeai.utils

import android.content.Context
import android.os.Build
import android.util.Log
import com.kreativekoala.scribeai.BuildConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Service for reporting user journey errors to the backend
 * Matches iOS ErrorReportingService implementation
 */
object ErrorReportingService {

    private const val TAG = "ErrorReportingService"

    /**
     * User journey flows that can be reported
     */
    enum class UserFlow(val displayName: String) {
        // Auth flows
        SIGN_IN("Sign In"),
        SIGN_UP("Sign Up"),
        SIGN_IN_WITH_GOOGLE("Sign In with Google"),
        SIGN_OUT("Sign Out"),

        // Content creation flows
        RECORD_AUDIO("Record Audio"),
        UPLOAD_PDF("Upload PDF"),
        SCAN_DOCUMENT("Scan Document"),
        YOUTUBE_LINK("YouTube Link"),
        UPLOAD_SLIDESHOW("Upload Slideshow"),

        // AI generation flows
        GENERATE_SUMMARY("Generate Summary"),
        GENERATE_QUIZ("Generate Quiz"),
        GENERATE_FLASHCARDS("Generate Flashcards"),
        GENERATE_PODCAST("Generate Podcast"),
        GENERATE_DIAGRAM("Generate Diagram"),
        GENERATE_INFOGRAPHIC("Generate Infographic"),
        CHAT_WITH_NOTE("Chat with Note"),

        // Subscription flows
        PURCHASE("Purchase Subscription"),
        RESTORE_PURCHASES("Restore Purchases"),

        // Other flows
        FETCH_NOTES("Fetch Notes"),
        DELETE_NOTE("Delete Note"),
        TRANSCRIPTION("Transcription")
    }

    // Dedicated coroutine scope for background error reporting
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    // Short-timeout client for error reporting
    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .build()

    private var authManager: AuthManager? = null
    private var appContext: Context? = null

    /**
     * Initialize the service with required dependencies
     * Should be called once during app startup
     */
    fun initialize(context: Context, authManager: AuthManager) {
        this.appContext = context.applicationContext
        this.authManager = authManager
    }

    /**
     * Report an error that occurred during a user journey
     * Runs completely in background - never blocks UI
     *
     * @param flow The user flow where the error occurred
     * @param error The exception that occurred
     * @param additionalInfo Any additional context
     */
    fun reportError(flow: UserFlow, error: Throwable, additionalInfo: String? = null) {
        val errorMessage = buildString {
            append(error.message ?: error.javaClass.simpleName)
            if (additionalInfo != null) {
                append(" | Additional: $additionalInfo")
            }
        }
        reportError(flow, errorMessage)
    }

    /**
     * Report an error with a custom error message
     * Runs completely in background - never blocks UI
     *
     * @param flow The user flow where the error occurred
     * @param errorMessage Custom error message
     * @param additionalInfo Any additional context
     */
    fun reportError(flow: UserFlow, errorMessage: String, additionalInfo: String? = null) {
        // Fire and forget - run in background
        scope.launch {
            try {
                sendErrorReport(flow, errorMessage, additionalInfo)
            } catch (e: Exception) {
                // Don't throw - just log locally
                Log.w(TAG, "Failed to send error report: ${e.message}")
            }
        }
    }

    private suspend fun sendErrorReport(flow: UserFlow, errorMessage: String, additionalInfo: String?) {
        val url = "${BuildConfig.BASE_URL.trimEnd('/')}/api/alerts/error"

        // Build full error message
        val fullErrorMessage = buildString {
            append(errorMessage)
            if (additionalInfo != null) {
                append(" | $additionalInfo")
            }
        }

        // Get device info
        val deviceInfo = JSONObject().apply {
            put("device", "${Build.MANUFACTURER} ${Build.MODEL}")
            put("osVersion", "Android ${Build.VERSION.RELEASE} (SDK ${Build.VERSION.SDK_INT})")
            put("appVersion", getAppVersion())
        }

        // Build request body
        val body = JSONObject().apply {
            put("flow", flow.displayName)
            put("error", fullErrorMessage)
            put("deviceInfo", deviceInfo)
        }

        // Build request
        val requestBuilder = Request.Builder()
            .url(url)
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .addHeader("Content-Type", "application/json")

        // Add auth token if available
        authManager?.getCurrentToken()?.let { token ->
            requestBuilder.addHeader("Authorization", "Bearer $token")
        }

        val request = requestBuilder.build()

        // Execute request
        client.newCall(request).execute().use { response ->
            if (response.isSuccessful) {
                Log.d(TAG, "📧 Error reported successfully for flow: ${flow.displayName}")
            } else {
                Log.w(TAG, "⚠️ Failed to report error, status: ${response.code}")
            }
        }
    }

    private fun getAppVersion(): String {
        return try {
            val context = appContext ?: return "Unknown"
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionName ?: "Unknown"
        } catch (e: Exception) {
            "Unknown"
        }
    }
}
